source/common/router/router.cc

```diff
@@ -110,6 +112,19 @@ bool convertRequestHeadersForInternalRedirect(Http::RequestHeaderMap& downstream
   downstream_headers.setHost(absolute_url.hostAndPort());
   downstream_headers.setPath(absolute_url.pathAndQueryParams());
 
+  callbacks->clearRouteCache();
+  auto route = callbacks->route();
+  // Don't allow a redirect to a non existing route.
+  if (!route) {
+    return false;
+  }
+
+  auto& route_name = route->routeEntry()->routeName();
+  for (auto& predicate : policy.predicates()) {
+    if (!predicate->acceptTargetRoute(*filter_state, route_name)) {
+      return false;
+    }
+  }
```

Comment:
These two cases are interesting, as they're the first two where we pass through the redirect while modifying the request headers.  I wonder if this is going to be confusing for logging since the request and response functionally no longer match, and if we should document this or actually latch and restore.  Thoughts?

Reply:
Added latching code.